### ISSUE TITLE
EOPNOTSUPP (Operation) not supported when using ACLs in non-privileged container (UID mapping + fuseoverlay-fs)

### ISSUE THREAD

**Issue by cdauth** [rfjakob/gocryptfs#640]
I have a host system running Fedora Server. On the host system I create an unprivileged LXD container (that uses UID mapping). When using gocryptfs inside the container, I get "Operation not supported" errors when trying to use ACLs. However, when using a privileged container (without UID mapping), I don't get these errors, also when not running as root.

I am not actually interested in using ACLs, but my use case is to run a Docker daemon with encrypted storage inside an LXD container. The `/var/lib/docker` directory should be encrypted with gocryptfs. Docker uses the fuse-overlayfs driver for its storage, and that driver tries to access the ACLs and fails. This results in strange behaviour inside Docker containers, where root cannot access files owned by other users, even if the permissions are right.

I am not sure, but I believe that normally even on file systems that don't support ACLs, `chacl -l` can be used to show the current file permissions without failing.

The problem can be reproduced inside an unprivileged LXD container or in a Docker Rootless container. In case it is interesting, using encfs leads to the same problem (see vgough/encfs#645). My detailed investigation of this problem is documented in the [LXD forum](https://discuss.linuxcontainers.org/t/not-supported-permission-error-when-using-fuse-docker-inside-unprivileged-container/13186).

# Steps to reproduce (LXD)

1. Install LXD
2. Start LXD: `systemctl start lxd`
3. Set up LXD: `lxd init` – the default settings should work
4. `echo root:1000000:1001000000 >> /etc/subuid`
5. `echo root:1000000:1001000000 >> /etc/subgid`
6. `lxc launch images:fedora/35 test`
7. `lxc exec test bash`

Now you are running a shell inside the LXD container. There, run the following commands:
```bash
dnf -y install gocryptfs acl
mkdir /mnt.enc
gocryptfs -init /mnt.enc
gocryptfs -allow_other /mnt.enc /mnt
touch /mnt/test
chacl -l /mnt/test
```

# Steps to reproduce (Docker Rootless)

When Docker is running in rootless mode, you can reproduce the problem by starting a Docker container using `docker run -ti --device /dev/fuse --cap-add SYS_ADMIN --rm fedora bash` and then running the same commands as in the LXD container above.

**Comment by cdauth** [comment]
I tested this now with encfs and encountered the same error. This indicates to me that the problem might lie somewhere else. I will continue digging.

**Comment by cdauth** [comment]
After some more digging, I understand that the error has to do with `fuse-overlayfs`, which is used by Docker as the storage driver. This is the error raised by `fuse-overlayfs` in debug mode:
```
unique: 20, opcode: GETXATTR (22), nodeid: 36608464, insize: 72, pid: 751
ovl_getxattr(ino=36608464, name=system.posix_acl_access, size=4096)
   unique: 20, error: -95 (Operation not supported), outsize: 16
```

The problem can be reproduced outside of Docker/fuse-overlayfs. On a gocryptfs filesystem inside an unprivileged LXD container, running `chacl -l test` shows an error `Operation not supported`. In a privileged LXD container however, both reading and setting the ACL works fine.

**Comment by rfjakob** [comment]
Hi, can you try the gocryptfs mount with added `-acl` flag?

**Comment by rfjakob** [comment]
Yep, it works with -acl, but you need [a newer gocryptfs](https://github.com/rfjakob/gocryptfs/releases/tag/v2.2.1) than what fedora has:

```
[root@test ~]# ./gocryptfs -version
gocryptfs v2.2.1 without_openssl; go-fuse v2.1.1-0.20210825171523-3ab5d95a30ae; 2021-10-20 go1.17.2 linux/amd64

[root@test ~]# ./gocryptfs -allow_other -acl /mnt.enc /mnt

[root@test ~]# chacl -l /mnt/test
/mnt/test [u::rw-,g::r--,o::r--]
```

**Comment by cdauth** [comment]
That's really interesting! I can confirm that it seems to work with that argument. I will see if that solves the problem with fuseoverlay-fs. I had no idea that this argument exists, it seems like it is neither in the man page nor in the help output.

I was not aware and it really sucks that Fedora ships with such an old version of gocryptfs. I'm quite fed up with Fedora’s limited selection of packages, but it seems like none of the other common distributions are any better. It seems like I will have to settle with Arch Linux in the end, even though it is maybe not the most stable distribution to run on a server.

**Comment by cdauth** [comment]
I can confirm that Docker works properly with fuse-overlayfs when the storage was mounted with `gocryptfs -allow_other -acl`. Thank you for your assistance!

I have also tested how other file systems without ACL support behave. In particular, I have mounted an ext4 file system with `-o noacl`. It actually seems to behave the same as gocryptfs, showing an “Operation not supported” error when trying to read the ACL of a file. Because of this, this whole situation actually seems like a bug in fuse-overlayfs, which I have reported there as containers/fuse-overlayfs#336.

One thing that I am still curious about is why it seems to work in a privileged container even without the `-acl` flag. Does gocryptfs somehow automatically enable ACL support when it detects that it is running as (non-UID-mapped) root?
